### ISSUE TITLE
[Discussion] Inferring Class Counts in Classifier Pipelines

### ISSUE THREAD

**Issue by RossKohler** [tensorflow/tfx#3776]
Is there any way for the pipeline to dynamically infer the number of classes that exist within the training/eval datasets? I feel like this would be best suited for the `StatisticsGen` component but I can't see any way I could get this working out of the box. Having the number of classes hardcoded in your code feels like it could be a source of error if you forget to modify it when new classes are added/removed. 

I ask because softmax layers expect you to define the number of unique classes in the definition of the model graph. I find I have to do some analysis outside of the pipeline to ensure that I get this number correct.

I could understand why this wouldn't be possible though. Since the model graph gets generated before the data gets ingested. Do any of the TFX developers have any thoughts or advice to provide on this subject?

**Comment by EdwardCuiPeacock** [comment]
Hi, @RossKohler . I was just exploring this a couple days ago. Here is the code I come up with after exploring their [interactive tfx component notebook(https://colab.research.google.com/github/tensorflow/tfx/blob/master/docs/tutorials/tfx/components.ipynb#scrollTo=FmPftrv0lEQy):

```
import os
import tensorflow_data_validation as tfdv
def get_unique_category_count(statistics_gen_output: dict, 
                              feature_name: str = "company", 
                              partition: str = "train") -> int:
    """
    Get the unique count of a string / categorical feature from StatisticsGen output
    Parameters:
    -----------
    * statistics_gen_output: dict
    	From statistics_gen.output["statistics"].to_json_dict(), where statistics_gen is 
        an instance of the TFX component StatisticsGen, usually constructed in some 
        `pipeline.py`
    * feature_name: str
    	Feature name
    * partition: str
    	Partition (train / test) of the dataset
        
    Returns:
    ----------
    * unique: int
    	A unqiue count of the string / categorical features
    """
    # Construct the protobuff file directory
    proto_path = os.path.join(statistics_gen_output["artifacts"][0]["artifact"]["uri"], f"Split-{partition}", "FeatureStats.pb")
    # Load the feature stats binary
    data_feature_list = tfdv.load_stats_binary(proto_path)
    # Get the stats from a specific feature
    feature_stats = tfdv.get_feature_stats(data_feature_list.datasets[0], tfdv.FeaturePath([feature_name]))
    # Get the string_stats (unique count, etc.) for this string feature. 
    # For other numerical features, use "num_stats" field
    unique_count = feature_stats.string_stats.unique
    return unique_count

# Example usage
get_unique_category_count(statistics_gen.outputs["statistics"].to_json_dict(), feature_name="company")
```
You may need to pass the output channel to the `custom_config` dictionary for Trainer or Transform component.

**Comment by gaikwadrahul8** [comment]
@RossKohler 

Apologies for the delay and I would request you to please have look into one Stack overflow solution and approach [here](https://stackoverflow.com/questions/65879049/multioutput-classification-with-tensorflow-extended-tfx) and even @EdwardCuiPeacock provided some workaround so please let us know, Is it resolving your issue?

TFMA also provides built-in support for converting binary classification metrics for use with multi-class/multi-label problems:

- Binarization based on class ID, top K, etc.
- Aggregated metrics based on micro averaging, macro averaging, etc.

TFMA also provides built-in support for query/ranking based metrics where the examples are grouped by a query key automatically in the pipeline.

Combined there are over 50+ standard metrics and plots available for a variety of problems including regression, binary classification, multi-class/multi-label classification, ranking, etc. so you can refer official documentation [here](https://www.tensorflow.org/tfx/model_analysis/metrics)

If you need any further assistance, please let us know ?

Thank you!

**Comment by gaikwadrahul8** [comment]
Hi, @RossKohler 

Closing this issue due to lack of recent activity for couple of weeks. Please feel free to reopen the issue, if you need any further assistance or if you have any questions.

Thank you!
